bw_timex/timex_lca.py

```diff
@@ -1053,35 +1055,89 @@ def create_node_id_collection_dict(self) -> None:
-                            warnings.warn(
-                                f"Failed to find process in database {background_db} for \
-                                    name='{exc.input['name']}', \
-                                    reference product='{exc.input['reference product']}', \
-                                    location='{exc.input['location']}': {e}"
-                            )
-                    self.interdatabase_activity_mapping.add(act_set)
 
         self.node_id_collection_dict["first_level_background_node_ids_static"] = (
             first_level_background_node_ids_static
         )
-        self.node_id_collection_dict["first_level_background_node_ids_all"] = (
-            first_level_background_node_ids_all
-        )
+   
+
+    def add_interdatabase_activity_mapping_from_timeline(self) -> None:
+        """
+        Fills the interdatabase_activity_mapping, which is a SetList of the matching processes across background databases
+        in the format of {(id, database_name_1), (id, database_name_2)} with only those activities and background databases that are actually mapped in the timeline.
+        Also adds the ids to the node_id_collection_dict["first_level_background_node_ids_interpolated"].
+        This avoids unneccessary peewee calls.
+
+        Parameters
+        ----------
+        None
+
+
+        Returns
+        -------
+        None, but adds the ids of producers in other background databases (only those interpolated to in the timeline) to the `interdatabase_activity_mapping` and `node_id_collection["first_level_background_node_ids_interpolated"]`.
+        """
+        if not hasattr(self, "timeline"):
+            warnings.warn("Timeline not yet built. Call TimexLCA.build_timeline() first.")
+            return
+                          
+        self.interdatabase_activity_mapping = SetList()
+        first_level_background_node_ids_interpolated = set()
+        unique_producer_background_db_combos = {}
+
+        # get unique combos of producers and background dbs from timeline
+        for _, row in self.timeline.iterrows():
+            if row["interpolation_weights"] is not None: # "None" corresponds to exchanges linked within the foreground -> skipped
+                
+                if row['producer'] not in unique_producer_background_db_combos.keys():
+                    unique_producer_background_db_combos[row['producer']] = set()
+                unique_producer_background_db_combos[row['producer']].update(row["interpolation_weights"].keys())
+
+        # fill interdatabase_activity_mapping with unique combos
+        for original_producer_id, background_databases in unique_producer_background_db_combos.items():
+            original_database = list(self.node_id_collection_dict["demand_dependent_background_database_names"])[0] #what if more than one orignal bg database is linked?
+            orig_act = bd.get_node(id=original_producer_id)
```

Comment:
If you feel like it's easy to do now, we could also save this information once at the very beginning and store it in a dict{database_id: node_object}. This could be useful at several other places as well. But maybe this would be a whole other PR --

Reply:
I would tackle this in a different pull request, e.g. when we sort out what exactly we want to store in the node_id_collection_dict.